### PR TITLE
RequestHedgingProxyProvider topic

### PR DESCRIPTION
**Pull request by benallamar**
Improvements:
      - Enhance the behavior of the proxy during many fail-overs:
      If the request is omitted before or during a failover the "toIgnore" blocks out
      with the value of the old active namenode. Thus, The patch force
      the proxy to check first with the last successful proxy or retry
      on the whole nodes again.

      - Enhance the logging: Add more information on the logging, and check
      the log level before calling either trace or debug

      - Fix tests: Adapt tests to match the current behavior of the proxy

### CODE REVIEW

File: hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/ha/RequestHedgingProxyProvider.java [hunk 1]
```diff
@@ -234,8 +234,6 @@ private void logProxyException(Exception ex, String proxyInfo) {
         LOG.debug("Invocation returned standby exception on [" +
             proxyInfo + "]");
       }
-    } else {
-      LOG.warn("Invocation returned exception on [" + proxyInfo + "]");
```
Comment (mk-dhia) [inline hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/ha/RequestHedgingProxyProvider.java]: why removing this message at least keep in debug , unless you are sure that when you reach this step you have already logged this exception elsewhere

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/ipc/Client.java [hunk 2]
```diff
@@ -854,7 +854,7 @@ private void handleConnectionFailure(int curRetries, IOException ioe
       if (action.action == RetryAction.RetryDecision.FAIL) {
         if (action.reason != null) {
           LOG.warn("Failed to connect to server: " + server + ": "
-              + action.reason, ioe);
+              + action.reason);
```
Comment (mk-dhia) [inline hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/ipc/Client.java:857]: Do we add the stack as debug if we want to undestand why we failed connecting?

Comment (benallamar) [reply hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/ipc/Client.java:857]: Yes! The logs for this part have been moved to the proxy level.

File: hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/ha/RequestHedgingProxyProvider.java (hunk 1 above)
Comment (benallamar) [reply hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/ha/RequestHedgingProxyProvider.java]: The logs are duplicated, they'll be shown at the level of RetryInvocationHandler caller.

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/retry/RetryPolicies.java [hunk 3]
```diff
@@ -260,18 +273,37 @@ protected long calculateSleepTime(int retries) {
...
+      this.maxTime = maxTime;
+      this.timeUnit = timeUnit;
+    }
+
+    @Override
+    protected String getReason() {
+      return constructReasonString(this.maxTime, this.timeUnit);
+    }
+
+    @VisibleForTesting
```
Comment (Willymontaz) [inline hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/retry/RetryPolicies.java:294]: I advice you not to test debug phrases, anoying to maintain and no great value.

File: hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/io/retry/TestRetryProxy.java [hunk 4]
```diff
@@ -18,59 +18,88 @@
...
   private UnreliableImplementation unreliableImpl;
+  private RetryAction caughtRetryAction = null;
   
   @Before
   public void setUp() throws Exception {
     unreliableImpl = new UnreliableImplementation();
   }
 
+  // answer mockPolicy's method with realPolicy, caught method's return value
+  private void setupMockPolicy(RetryPolicy mockPolicy,
```
Comment (Willymontaz) [inline hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/io/retry/TestRetryProxy.java:60]: If I remember Mockito Spy does what you want in a shorter and cleaner fashion (verify on a real object and not a mock)

File: hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/io/retry/TestRetryProxy.java [hunk 5]
```diff
@@ -80,25 +109,21 @@ public void testRpcInvocation() throws Exception {
...
-    
+
+    final AtomicInteger count = new AtomicInteger();
     // Embed the proxy in ProtocolTranslator
     ProtocolTranslator xlator = new ProtocolTranslator() {
-      int count = 0;
       @Override
       public Object getUnderlyingProxyObject() {
-        count++;
+        count.getAndIncrement();
```
Comment (Willymontaz) [inline hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/io/retry/TestRetryProxy.java:118]: changes on this test are not part of this Pull Request

File: hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/ha/RequestHedgingProxyProvider.java [hunk 6]
```diff
@@ -140,14 +150,36 @@ public Object call() throws Exception {
         }
       }
     }
-  }
 
+    /**
+     * Get the successful/Left proxy
+     * If a proxy has failed, we, either, return the successful one, or the one not yet tested.
```
Comment (Willymontaz) [inline hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/ha/RequestHedgingProxyProvider.java:158]: This comment sounds confusing -> "if the proxy has failed we return the successful one" that does not fit with the fact that this method is called at the very beginning even before trying any proxy at all

Comment (benallamar) [reply hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/ha/RequestHedgingProxyProvider.java:158]: Done!

**Comment by Willymontaz** [comment]
LGTM
